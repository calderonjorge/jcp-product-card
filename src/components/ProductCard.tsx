import React , { createContext, useContext, CSSProperties } from 'react';
import { useProduct } from '../hooks/useProduct';

import { ProductContextProps, Product, onChangeArgs, InitialValues, ProductCardHandlers } from '../interfaces/Product.interfaces';
import styles from '../styles/styles.module.css';

const ProductContext = createContext<ProductContextProps>({} as ProductContextProps);

export const useProductContext = () => useContext(ProductContext);

export interface Props {
    product: Product,
    children: (args: ProductCardHandlers) => JSX.Element,
    initialValues?: InitialValues,
    className?: string,
    style?: CSSProperties,
    value?: number,
    onChange?: (args: onChangeArgs) => void,
}

export const ProductCard = ({ product, children, className, style, value, onChange, initialValues }: Props) => {
    
    const { counter, increaseBy, maxCount, isMaxCountReached, reset } = useProduct({ onChange, product, value, initialValues });


    return (
        <ProductContext.Provider value={{ counter, increaseBy, product, maxCount, isMaxCountReached }}>
            <div className={`${styles.productCard} ${className}`} style={style}>
                {children(
                    {
                        count: counter,
                        increaseBy,
                        isMaxCountReached,
                        maxCount,
                        product,
                        reset,
                    }
                )}
            </div>
        </ProductContext.Provider>
    )
}
