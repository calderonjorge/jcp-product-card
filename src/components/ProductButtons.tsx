import React from 'react';
import { useProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css';
import { CSSProperties } from "react";

export interface Props {
    maxCount?: number;
    className?: string;
    style?: CSSProperties;
}

export const ProductButtons = ({ className, style }: Props) => {

    const { increaseBy, counter, isMaxCountReached } = useProductContext();
    
    return (
        <div className={`${styles.buttonsContainer} ${className}`} style={style}>
            <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>-</button>
            <div className={styles.countLabel}>{counter}</div>
            <button
                className={`${styles.buttonAdd} ${isMaxCountReached && styles.disabled}`}
                onClick={() => { increaseBy(1) }}
                disabled={isMaxCountReached}
            >
                +
            </button>
        </div>
    )
}