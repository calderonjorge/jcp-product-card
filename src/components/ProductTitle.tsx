import React from 'react';
import { useProductContext } from "./ProductCard";
import styles from '../styles/styles.module.css'
import { CSSProperties } from "react";

export interface Props {
    className?: string
    title?: string,
    style?: CSSProperties
}

export const ProductTitle = ({ title, className, style }: Props) => {
    const { product } = useProductContext();

    return <span className={`${styles.productDescription} ${className}`} style={style}>{title ? title : product.title}</span>
}