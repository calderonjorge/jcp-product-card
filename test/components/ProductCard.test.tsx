import React from 'react';
import rendered from 'react-test-renderer'
import { ProductCard } from '../../src/components';
import { product2 } from '../data/products';

const { act } = rendered;

describe('Product Card', () => {

    test('should the component', () => {
        const wraper = rendered.create(
            <ProductCard product={product2}>
                {
                    () => (
                        <h1>Product Card</h1>
                    )
                }
            </ProductCard>
        );
        expect(wraper.toJSON()).toMatchSnapshot();
    });



    test('should increse the counter', () => {
        const wraper = rendered.create(
            <ProductCard product={product2}>
                {
                    ({ count, increaseBy }) => (
                        <>
                            <h1>Product Card</h1>
                            <span>{count}</span>
                            <button onClick={() => { increaseBy(1) }}>Click Me</button>
                        </>
                    )
                }
            </ProductCard>
        );
        let tree = wraper.toJSON();
        expect(tree).toMatchSnapshot();
        act(()=>{
            (tree as any).children[2].props.onClick();
        })
        tree = wraper.toJSON();
        expect((tree as any).children[1].children[0]).toBe('1');
    });
});
