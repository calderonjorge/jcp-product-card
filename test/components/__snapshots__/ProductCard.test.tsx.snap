// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Card should increse the counter 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    Click Me
  </button>
</div>
`;

exports[`Product Card should the component 1`] = `
<div
  className="productCard undefined"
>
  <h1>
    Product Card
  </h1>
</div>
`;
